File: src/erlxc.erl
%%% Copyright (c) 2013-2014, Michael Santos <michael.santos@gmail.com>
%%% Permission to use, copy, modify, and/or distribute this software for any
%%% purpose with or without fee is hereby granted, provided that the above
%%% copyright notice and this permission notice appear in all copies.
%%%
%%% THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
%%% WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
%%% MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
%%% ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
%%% WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
%%% ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
%%% OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
-module(erlxc).
-include_lib("erlxc/include/erlxc.hrl").

-export([
        spawn/0, spawn/1, spawn/2,

        type/1,
        temporary/1, transitory/1, permanent/1,

        send/2,
        exit/2,

        container/1,
        console/1
    ]).

-type container() :: #container{port::port(),console::port()}.

-spec spawn() -> container().
-spec spawn(string() | binary()) -> container().
-spec spawn(string() | binary(),list()) -> container().
spawn() ->
    erlxc:spawn(<<>>, []).
spawn(Name) ->
    erlxc:spawn(Name, []).
spawn(<<>>, Options) ->
    erlxc:spawn(name(<<"erlxc">>), Options ++ [temporary]);
spawn(Name, Options) ->
    Port = erlxc_drv:start(Name, Options ++ [transitory]),
    boot(#container{port = Port}, Options).

-spec send(container(),iodata()) -> 'true'.
send(#container{console = Console}, Data) ->
    erlxc_console:send(Console, Data).

-spec exit(container(),'kill' | 'normal') -> boolean().
exit(#container{port = Port}, normal) ->
    liblxc:shutdown(Port, 0);

exit(#container{port = Port}, kill) ->
    liblxc:stop(Port).

type(#container{port = Port}) ->
    liblxc:type(Port).

temporary(#container{port = Port}) ->
    liblxc:temporary(Port).

transitory(#container{port = Port}) ->
    liblxc:transitory(Port).

permanent(#container{port = Port}) ->
    liblxc:permanent(Port).

container(#container{port = Port}) -> Port.
console(#container{console = Port}) -> Port.

%%--------------------------------------------------------------------
%%% Container state
%%--------------------------------------------------------------------

% "STOPPED", "STARTING", "RUNNING", "STOPPING",
% "ABORTING", "FREEZING", "FROZEN", "THAWED",
boot(#container{port = Port} = Container, Options) ->
    true = liblxc:async_state_notify(Port, 5),
    state(Container, Options).

state(#container{port = Port} = Container, Options) ->
    state(Container, liblxc:defined(Port), Options).
state(#container{port = Port} = Container, false, Options) ->
    case erlxc_drv:event(Port, infinity) of
        {state, <<"STOPPED">>} ->
            create(Container, Options),
            config(Container, Options),
            start(Container, Options),
            state(Container, Options);
        {state, State} ->
            erlang:error({error, State})
    end;
state(#container{port = Port} = Container, true, Options) ->
    case erlxc_drv:event(Port, infinity) of
        {state, <<"RUNNING">>} ->
            true = liblxc:async_state_close(Port),
            Name = liblxc:name(Port),
            Console = erlxc_console:start(Name),
            Container#container{console = Console};
        {state, <<"STOPPED">>} ->
            config(Container, Options),
            start(Container, Options),
            state(Container, Options);
        {state, <<"FROZEN">>} ->
            true = liblxc:unfreeze(Port),
            state(Container, true, Options);
        {state, State} when State =:= <<"STARTING">>; State =:= <<"STOPPING">>;
                State =:= <<"FREEZING">>; State =:= <<"THAWED">> ->
            state(Container, Options);
        {state, State} ->
            erlang:error({unsupported, State, Options})
    end.

config(#container{port = Port}, Options) ->
    Config = proplists:get_value(config, Options, []),

    [ begin
        case Item of
            <<>> ->
                true = liblxc:clear_config(Port);
            {Key, Value} ->
                true = liblxc:set_config_item(Port, Key, Value);
            Key ->
                true = liblxc:clear_config_item(Port, Key)
        end
      end || Item <- Config ].

create(#container{port = Port}, Options) ->
    Create = proplists:get_value(create, Options, []),

    Template = proplists:get_value(template, Create, <<"ubuntu">>),
    Bdevtype = proplists:get_value(bdevtype, Create, <<>>),
    Bdevspec = proplists:get_value(bdevspec, Create, <<>>),
    Flags = proplists:get_value(flags, Create, 0),
    Argv = proplists:get_value(argv, Create, []),

    true = liblxc:create(Port, Template, Bdevtype, Bdevspec, Flags, Argv).

start(#container{port = Port}, Options) ->
    Start = proplists:get_value(start, Options, []),
    UseInit = proplists:get_value(useinit, Start, false),
    Argv = proplists:get_value(argv, Start, []),

    true = liblxc:start(Port, bool(UseInit), Argv).

%%--------------------------------------------------------------------
%%% Internal functions
%%--------------------------------------------------------------------
bool(true) -> 1;
bool(false) -> 0;
bool(1) -> 1;
bool(0) -> 0.

i2b(N) ->
    list_to_binary(integer_to_list(N)).

name(Name) ->
    % XXX possible to re-use container names
    N = binary:decode_unsigned(crypto:rand_bytes(4)),
    <<Name/binary, (i2b(N))/binary>>.

File: c_src/erlxc.c
/* Copyright (c) 2013-2014, Michael Santos <michael.santos@gmail.com>
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */
#include "erlxc.h"
#include "erlxc_version.h"

#define ERLXC_MSG_SYNC  0
#define ERLXC_MSG_ASYNC (htons(1))

static void erlxc_loop(erlxc_state_t *);
static void usage(erlxc_state_t *);

static int erlxc_write(u_int16_t, ETERM *);
static ssize_t erlxc_read(void *, ssize_t);

static void erlxc_stats(erlxc_state_t *ep);

extern char *__progname;

    int
main(int argc, char *argv[])
{
    erlxc_state_t *ep = NULL;
    char *name = NULL;
    char *path = NULL;
    char *errlog = NULL;
    int ch = 0;

    erl_init(NULL, 0);

    ep = calloc(1, sizeof(erlxc_state_t));
    if (!ep)
        erl_err_sys("calloc");

    ep->opt |= erlxc_opt_stop_on_exit;
    ep->opt |= erlxc_opt_daemonize;
    ep->opt |= erlxc_opt_closeallfds;

    while ( (ch = getopt(argc, argv, "d:hn:o:P:t:v")) != -1) {
        switch (ch) {
            case 'd':
                if (strcmp("nodaemonize", optarg) == 0)
                    ep->opt &= ~erlxc_opt_daemonize;
                else if (strcmp("nocloseallfds", optarg) == 0)
                    ep->opt &= ~erlxc_opt_closeallfds;
                break;
            case 'n':
                name = strdup(optarg);
                if (!name)
                    erl_err_sys("name");
                break;
            case 'o':
                errlog = strdup(optarg);
                if (!errlog)
                    erl_err_sys("errlog");
                break;
            case 'P':
                path = strdup(optarg);
                if (!path)
                    erl_err_sys("path");
                break;
            case 't':
                if (strcmp("temporary", optarg) == 0) {
                    ep->opt |= erlxc_opt_stop_on_exit;
                    ep->opt |= erlxc_opt_destroy_on_exit;
                }
                else if (strcmp("transient", optarg) == 0) {
                    ep->opt |= erlxc_opt_stop_on_exit;
                    ep->opt &= ~erlxc_opt_destroy_on_exit;
                }
                else if (strcmp("permanent", optarg) == 0) {
                    ep->opt &= ~erlxc_opt_stop_on_exit;
                    ep->opt &= ~erlxc_opt_destroy_on_exit;
                }
                else
                    usage(ep);
                break;
            case 'v':
                ep->verbose++;
                break;
            case 'h':
            default:
                usage(ep);
        }
    }

    if (!name)
        usage(ep);

    if (errlog) {
        if (!freopen(errlog, "w+", stderr))
            erl_err_sys("freopen");
    }

    ep->c = lxc_container_new(name, path);
    if (!ep->c)
        erl_err_quit("failed to create container");

    free(name);
    free(path);

    erlxc_loop(ep);

    if (ep->opt & erlxc_opt_stop_on_exit) {
        VERBOSE(1, "stopping container:%s", ep->c->name);
        (void)ep->c->stop(ep->c);
        (void)ep->c->wait(ep->c, "STOPPED", -1);
    }

    if (ep->opt & erlxc_opt_destroy_on_exit) {
        VERBOSE(1, "destroying container:%s", ep->c->name);
        (void)ep->c->destroy(ep->c);
    }

    exit(0);
}

    static void
erlxc_loop(erlxc_state_t *ep)
{
    unsigned char buf[UINT16_MAX] = {0};
    unsigned char *msg = buf;
    ETERM *arg = NULL;
    ETERM *reply = NULL;
    u_int16_t len = 0;
    u_int16_t cmd = 0;

    for ( ; ; ) {
        if (erlxc_read(&len, sizeof(len)) != sizeof(len))
            return;

        len = ntohs(len);

        if (len <= sizeof(cmd) || len > sizeof(buf))
            return;

        if (erlxc_read(msg, len) != len)
            return;

        cmd = get_int16(msg);
        msg += 2;
        len -= 2;

        arg = erl_decode(msg);
        if (!arg)
            erl_err_quit("invalid message");

        reply = erlxc_cmd(ep, cmd, arg);
        if (!reply)
            erl_err_quit("unrecoverable error");

        erl_free_compound(arg);

        if (erlxc_write(ERLXC_MSG_SYNC, reply) < 0)
            erl_err_sys("erlxc_write");

        erl_free_compound(reply);

        if (ep->verbose > 1)
            erlxc_stats(ep);

        (void)fflush(stderr);
    }
}

    int
erlxc_send(ETERM *t)
{
    return erlxc_write(ERLXC_MSG_ASYNC, t);
}

    static int
erlxc_write(u_int16_t type, ETERM *t)
{
    struct iovec iov[3];

    u_int16_t len;
    unsigned char buf[UINT16_MAX-4];
    int buflen = 0;

    buflen = erl_term_len(t);
    if (buflen < 0 || buflen > sizeof(buf))
        return -1;

    if (erl_encode(t, buf) < 1)
        return -1;

    len = htons(sizeof(type) + buflen);

    iov[0].iov_base = &len;
    iov[0].iov_len = sizeof(len);
    iov[1].iov_base = &type;
    iov[1].iov_len = sizeof(type);
    iov[2].iov_base = buf;
    iov[2].iov_len = buflen;

    return writev(STDOUT_FILENO, iov, sizeof(iov)/sizeof(iov[0]));
}

    static ssize_t
erlxc_read(void *buf, ssize_t len)
{
    ssize_t i = 0;
    ssize_t got = 0;

    do {
        if ((i = read(STDIN_FILENO, buf + got, len - got)) <= 0)
            return(i);
        got += i;
    } while (got < len);

    return len;
}

    static void
erlxc_stats(erlxc_state_t *ep)
{
    unsigned long allocated = 0;
    unsigned long freed = 0;

    erl_eterm_statistics(&allocated, &freed);
    VERBOSE(0, "allocated=%ld, freed=%ld", allocated, freed);
    erl_eterm_release();
}

    static void
usage(erlxc_state_t *ep)
{
    (void)fprintf(stderr, "%s %s (lxc %s)\n",
            __progname, ERLXC_VERSION, lxc_get_version());
    (void)fprintf(stderr,
            "usage: %s -n <name> <options>\n"
            "    -n <name>        container name\n"
            "    -o <path>        error log\n"
            "    -P <path>        LXC path\n"
            "    -v               verbose mode\n"
            "    -d <option>      debug: nodaemonize, nocloseallfds\n"
            "    -t <type>        container type (permanent, transient, temporary)\n",
            __progname
            );

    exit (EXIT_FAILURE);
}
